<launch>
    <!--  ************** Global Parameters ***************  -->
    <param name="/use_sim_time" value="true"/>
    <arg name="controller"  default="mpc" doc="opt: dwa, mpc, pure_pursuit"/> 
    <arg name="map"  default="empty" doc="opt: empty"/> 

    <arg name="r1_x_pos" default="0.0"/>
    <arg name="r1_y_pos" default="0.0"/>
    <arg name="r1_z_pos" default="0.0"/>
    <arg name="r1_roll" default="0.0"/>
    <arg name="r1_pitch" default="0.0"/>
    <arg name="r1_yaw" default="0.0"/>

    <arg name="r2_x_pos" default="0.0"/>
    <arg name="r2_y_pos" default="-2.0"/>
    <arg name="r2_z_pos" default="0.0"/>
    <arg name="r2_roll" default="0.0"/>
    <arg name="r2_pitch" default="0.0"/>
    <arg name="r2_yaw" default="0.0"/>

    <arg name="r3_x_pos" default="0.0"/>
    <arg name="r3_y_pos" default="2.0"/>
    <arg name="r3_z_pos" default="0.0"/>
    <arg name="r3_roll" default="0.0"/>
    <arg name="r3_pitch" default="0.0"/>
    <arg name="r3_yaw" default="0.0"/>

    <arg name="use_map_topic"   default="false"/>
    <arg name="scan_topic"      default="/robot_0/scan"/>
    <arg name="initial_pose_x"  default="3.0"/>
    <arg name="initial_pose_y"  default="1.0"/>
    <arg name="initial_pose_a"  default="0.0"/>
    <arg name="odom_frame_id"   default="robot_0/odom"/>
    <arg name="base_frame_id"   default="robot_0/base_footprint"/>
    <arg name="global_frame_id" default="map"/>

    <arg name="robot1"  default="robot_0"/>

    <!--  ************** Map Server **************  -->
    <node name="map_server" pkg="map_server" type="map_server" args="$(find gazebo_rl_test)/map/simul_map.yaml" output="screen" if="$(eval map == 'empty')">
        <param name="frame_id" value="/map"/>
    </node>
    -
    <!--  ************** Localization **************  -->
    <node pkg="amcl" type="amcl" name="robot_0_amcl" output="screen">
        <param name="use_map_topic"             value="$(arg use_map_topic)"/>
        <!-- Publish scans from best pose at a max of 10 Hz -->
        <param name="odom_model_type"           value="diff"/>
        <param name="odom_alpha5"               value="0.1"/>
        <param name="gui_publish_rate"          value="10.0"/>
        <param name="laser_max_beams"             value="60"/>
        <param name="laser_max_range"           value="12.0"/>
        <param name="min_particles"             value="500"/>
        <param name="max_particles"             value="2000"/>
        <param name="kld_err"                   value="0.05"/>
        <param name="kld_z"                     value="0.99"/>
        <param name="odom_alpha1"               value="0.2"/>
        <param name="odom_alpha2"               value="0.2"/>
        <!-- translation std dev, m -->
        <param name="odom_alpha3"               value="0.2"/>
        <param name="odom_alpha4"               value="0.2"/>
        <param name="laser_z_hit"               value="0.5"/>
        <param name="laser_z_short"             value="0.05"/>
        <param name="laser_z_max"               value="0.05"/>
        <param name="laser_z_rand"              value="0.5"/>
        <param name="laser_sigma_hit"           value="0.2"/>
        <param name="laser_lambda_short"        value="0.1"/>
        <param name="laser_model_type"          value="likelihood_field"/>
        <!-- <param name="laser_model_type" value="beam"/> -->
        <param name="laser_likelihood_max_dist" value="2.0"/>
        <param name="update_min_d"              value="0.25"/>
        <param name="update_min_a"              value="0.2"/>
        <param name="odom_frame_id"             value="$(arg odom_frame_id)"/>
        <param name="base_frame_id"             value="$(arg base_frame_id)"/>
        <param name="global_frame_id"           value="$(arg global_frame_id)"/>
        <param name="resample_interval"         value="1"/>
        <!-- Increase tolerance because the computer can get quite busy -->
        <param name="transform_tolerance"       value="1.0"/>
        <param name="recovery_alpha_slow"       value="0.0"/>
        <param name="recovery_alpha_fast"       value="0.0"/>
        <param name="initial_pose_x"            value="$(arg initial_pose_x)"/>
        <param name="initial_pose_y"            value="$(arg initial_pose_y)"/>
        <param name="initial_pose_a"            value="$(arg initial_pose_a)"/>
        <remap from="scan"                      to="$(arg scan_topic)"/>
        <remap from="initialpose"               to="/robot_0/initialpose"/>
        <remap from="amcl_pose"                 to="/robot_0/amcl_pose"/>
        <remap from="particlecloud"             to="/robot_0/particlecloud"/>
    </node>
    
    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
        <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
        <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/carlike/2costmap_common_params.yaml" command="load" ns="global_costmap" />
        <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/carlike/2costmap_common_params.yaml" command="load" ns="local_costmap" />
        <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/carlike/2local_costmap_params.yaml" command="load" />
        <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/carlike/2global_costmap_params.yaml" command="load" />      
        <param name="global_costmap/robot_base_frame" value="robot_0/base_footprint"/>
        <param name="local_costmap/robot_base_frame" value="robot_0/base_footprint"/>
        <param name="local_costmap/global_frame" value="robot_0/odom"/>
        <!-- remap from="map" to="/map" / -->
        <!-- remap from="cmd_vel" to="$(arg cmd_vel_topic)"/ -->
        <!-- remap from="odom" to="$(arg odom_topic)"/ -->
    </node>
    <!--  ************** Visualisation **************  -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find gazebo_rl_test)/launch/Simulation/params/rviz_rl_navigation.rviz"/>
</launch>
